runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java

```diff
@@ -0,0 +1,382 @@
+    ExpansionApi.ExpansionRequest request =
+        requestBuilder
+            .setComponents(runnerAPIpipeline.getComponents())
+            .setTransform(ptransformBuilder.build())
+            .setNamespace(UPGRADE_NAMESPACE)
+            .build();
+
+    ExpansionApi.ExpansionResponse response =
+        clientFactory.getExpansionServiceClient(transformServiceEndpoint).expand(request);
+
+    if (!Strings.isNullOrEmpty(response.getError())) {
+      throw new IOException(String.format("expansion service error: %s", response.getError()));
+    }
+
+    Map<String, RunnerApi.Environment> newEnvironmentsWithDependencies =
+        response.getComponents().getEnvironmentsMap().entrySet().stream()
+            .filter(
+                kv ->
+                    !runnerAPIpipeline.getComponents().getEnvironmentsMap().containsKey(kv.getKey())
+                        && kv.getValue().getDependenciesCount() != 0)
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    RunnerApi.Components expandedComponents =
+        response
+            .getComponents()
+            .toBuilder()
+            .putAllEnvironments(
+                External.ExpandableTransform.resolveArtifacts(
+                    newEnvironmentsWithDependencies, transformServiceEndpoint))
+            .build();
+    RunnerApi.PTransform expandedTransform = response.getTransform();
+    List<String> expandedRequirements = response.getRequirementsList();
+
+    RunnerApi.Components.Builder newComponentsBuilder = expandedComponents.toBuilder();
+
+    // Some transforms may refer to already overridden transform as one of their input. We record
+    // such occurrences and correct them by referring to the upgraded transform instead.
+    Collection<String> oldOutputs = transformToUpgrade.getOutputsMap().values();
+    Map<String, String> inputReplacements = new HashMap<>();
+    if (transformToUpgrade.getOutputsMap().size() == 1) {
+      inputReplacements.put(
+          oldOutputs.iterator().next(),
+          expandedTransform.getOutputsMap().values().iterator().next());
+    } else {
+      for (Map.Entry<String, String> entry : transformToUpgrade.getOutputsMap().entrySet()) {
+        if (expandedTransform.getOutputsMap().keySet().contains(entry.getKey())) {
+          throw new Exception(
+              "Original transform did not have an output with tag "
+                  + entry.getKey()
+                  + " but upgraded transform did.");
+        }
+        String newOutput = expandedTransform.getOutputsMap().get(entry.getKey());
+        if (newOutput == null) {
+          throw new Exception(
+              "Could not find an output with tag "
+                  + entry.getKey()
+                  + " for the transform "
+                  + expandedTransform);
+        }
+        inputReplacements.put(entry.getValue(), newOutput);
+      }
+    }
+
+    String newTransformId = transformId + "_upgraded";
```

Comment:
We can't guarantee this won't be a collision. On the other hand, there's no need to give it a new name at all, just put the new transform in the map at the old transforms name (and all references will remain valid).

Reply:
Done.